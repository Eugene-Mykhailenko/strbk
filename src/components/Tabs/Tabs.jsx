import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import styles from './Tabs.scss';

const Tabs = (props) => {
  const { wide, activeTabId, tabs, onClick } = props;

  const getClassNames = (tab) =>
    classNames(styles.tab, {
      [styles.active]: tab.id === activeTabId,
    });

  const renderTabs = () =>
    tabs.map((tab) => (
      <div
        className={getClassNames(tab)}
        key={tab.id}
        onClick={() => {
          onClick(tab.id);
        }}
      >
        {tab.title}
      </div>
    ));

  return (
    <div
      className={classNames(styles.tabs, {
        [styles.wide]: wide,
      })}
    >
      {renderTabs(props)}
    </div>
  );
};

Tabs.styles = styles;

Tabs.propTypes = {
  /**  Items list */
  tabs: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      title: PropTypes.string,
    }),
  ),
  /** This is when the tabs occupy the entire width of the lines. */
  wide: PropTypes.bool,
  /** It's an action when the user clicks on the tab. */
  onClick: PropTypes.func.isRequired,
  /** It's an ID of selected tab. */
  activeTabId: PropTypes.number,
};

Tabs.defaultProps = {
  tabs: [],
  wide: false,
  activeTabId: undefined,
};

export default Tabs;
