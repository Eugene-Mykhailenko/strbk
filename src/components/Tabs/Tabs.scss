@import '../../assets/styles/abstracts/variables';

.tabs {
  display: flex;
  white-space: nowrap;
  background-color: var(--color_9);

  .tab {
    display: inline;
    white-space: nowrap;
    overflow: hidden;
    padding: 13px 10px;
    border-bottom: 2px solid var(--color_8);
    text-transform: uppercase;
    font-weight: 400;
    font-size: 14px;
    text-overflow: ellipsis;
    color: var(--color_6);
    transition: $transitionColor, $transitionBorderColor;
    cursor: pointer;

    &.active {
      border-bottom: 2px solid var(--color_1);
      color: var(--color_3);
    }
  }

  &.wide {
    .tab {
      display: flex;
      justify-content: center;
      flex: 1;
    }
  }
}
