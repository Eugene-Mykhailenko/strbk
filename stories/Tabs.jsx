import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { boolean } from '@storybook/addon-knobs';
import { withState } from '@dump247/storybook-state';

import Tabs from '../src/components/Tabs/Tabs';
import storyStyles from './Stories.scss';

const addState = withState({ activeTabId: 1 });
const addInfo = withInfo({
  text: `Tabs component: a single content area with multiple panels, 
  each associated with a header in a list.`,
  inline: true,
});
const props = {
  tabs: [
    {
      id: 1,
      title: 'Top Matches',
    },
    {
      id: 2,
      title: 'Soon in LIVE',
    },
    {
      id: 3,
      title: 'Express',
    },
    {
      id: 4,
      title: 'Ordinary',
    },
  ],
};

/* eslint-disable-next-line react/prop-types */
const wrapper = ({ store }) => (
  <div className={storyStyles.container}>
    <div className={storyStyles.col}>
      <Tabs
        {...store.state}
        {...props}
        wide={boolean('tab-wide', false)}
        onClick={(activeTabId) => store.set({ activeTabId })}
      />
    </div>
  </div>
);

storiesOf('Tabs', module).add('Tabs', addState(addInfo(wrapper)));
