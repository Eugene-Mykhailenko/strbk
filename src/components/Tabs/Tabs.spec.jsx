import React from 'react';
import Tabs from './Tabs';

const config = {
  tabs: [
    {
      id: 1,
      title: 'Top Matches',
    },
    {
      id: 2,
      title: 'Soon in LIVE',
    },
    {
      id: 3,
      title: 'Express',
    },
    {
      id: 4,
      title: 'Ordinary',
    },
  ],
  onClick: jest.fn(),
};

const tabIds = [1, 2, 3, 4];

describe('Tabs', () => {
  it('should render with config', () => {
    const tabs = shallow(<Tabs {...config} />);

    expect(tabs.children().length).toBe(4);

    expect(tabs).toMatchSnapshot();
  });

  it('should pass proper configuration', () => {
    const tabs = shallow(<Tabs {...config} />);

    config.tabs.forEach((item, index) => {
      expect(tabs.props().children[index].key).toEqual(
        tabIds[index].toString(),
      );
    });

    expect(tabs.childAt(0).text()).toEqual(config.tabs[0].title);
  });

  it('should highlight active tab', () => {
    const tabs = shallow(<Tabs activeTabId={1} {...config} />);

    expect(tabs.childAt(0).prop('className')).toMatch(/active/);

    expect(tabs).toMatchSnapshot();
  });

  it('should be able to render wide tab', () => {
    const tabs = shallow(<Tabs wide {...config} />);

    expect(tabs.prop('className')).toMatch(/wide/);

    expect(tabs).toMatchSnapshot();
  });

  it('should handle click event', () => {
    const tabs = mount(<Tabs {...config} />);

    tabs
      .childAt(0)
      .childAt(1)
      .find('.tab')
      .simulate('click');

    expect(config.onClick).toHaveBeenCalledTimes(1);
    expect(config.onClick).toHaveBeenCalledWith(2);

    config.onClick.mockClear();

    expect(tabs).toMatchSnapshot();
  });
});
