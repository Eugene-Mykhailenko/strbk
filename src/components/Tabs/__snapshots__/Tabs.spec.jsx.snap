// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should be able to render wide tab 1`] = `
<div
  className="tabs wide"
>
  <div
    className="tab"
    key="1"
    onClick={[Function]}
  >
    Top Matches
  </div>
  <div
    className="tab"
    key="2"
    onClick={[Function]}
  >
    Soon in LIVE
  </div>
  <div
    className="tab"
    key="3"
    onClick={[Function]}
  >
    Express
  </div>
  <div
    className="tab"
    key="4"
    onClick={[Function]}
  >
    Ordinary
  </div>
</div>
`;

exports[`Tabs should handle click event 1`] = `
<Tabs
  onClick={[MockFunction]}
  tabs={
    Array [
      Object {
        "id": 1,
        "title": "Top Matches",
      },
      Object {
        "id": 2,
        "title": "Soon in LIVE",
      },
      Object {
        "id": 3,
        "title": "Express",
      },
      Object {
        "id": 4,
        "title": "Ordinary",
      },
    ]
  }
  wide={false}
>
  <div
    className="tabs"
  >
    <div
      className="tab"
      key="1"
      onClick={[Function]}
    >
      Top Matches
    </div>
    <div
      className="tab"
      key="2"
      onClick={[Function]}
    >
      Soon in LIVE
    </div>
    <div
      className="tab"
      key="3"
      onClick={[Function]}
    >
      Express
    </div>
    <div
      className="tab"
      key="4"
      onClick={[Function]}
    >
      Ordinary
    </div>
  </div>
</Tabs>
`;

exports[`Tabs should highlight active tab 1`] = `
<div
  className="tabs"
>
  <div
    className="tab active"
    key="1"
    onClick={[Function]}
  >
    Top Matches
  </div>
  <div
    className="tab"
    key="2"
    onClick={[Function]}
  >
    Soon in LIVE
  </div>
  <div
    className="tab"
    key="3"
    onClick={[Function]}
  >
    Express
  </div>
  <div
    className="tab"
    key="4"
    onClick={[Function]}
  >
    Ordinary
  </div>
</div>
`;

exports[`Tabs should render with config 1`] = `
<div
  className="tabs"
>
  <div
    className="tab"
    key="1"
    onClick={[Function]}
  >
    Top Matches
  </div>
  <div
    className="tab"
    key="2"
    onClick={[Function]}
  >
    Soon in LIVE
  </div>
  <div
    className="tab"
    key="3"
    onClick={[Function]}
  >
    Express
  </div>
  <div
    className="tab"
    key="4"
    onClick={[Function]}
  >
    Ordinary
  </div>
</div>
`;
